### PR TITLE
To solve the misisng column name uniqueness case we concatenate the c…

### PR DESCRIPTION
**Pull request by JohnnyC08**
Addresses https://github.com/scikit-learn-contrib/categorical-encoding/issues/132

We maintain a map of the number of times we see the raw column and append the `#` the number of times we have seen that column. By using a map and updating and using it one loop through everything, we hope we don't add to the overall processing time.

### CODE REVIEW

**Comment by janmotl** [comment]
Beware that we can still get a clash of names:
```python
    def test_one_hot_duplicate_names(self):
        X = pd.DataFrame({'sea_rch': ['dummy', 'found'],
                          'sea': ['rch#-1', 'found']})
        encoder = encoders.OneHotEncoder(use_cat_names=True)

        result = encoder.fit_transform(X)
        columns = result.columns.tolist()
        
        self.assertEqual(len(set(columns)), len(columns))
```

A citation that came to my mind:
> There are only two hard things in Computer Science: cache invalidation and naming things. -- Phil Karlton

What we are trying to deal with is an old problem. In Lisp they have gensym. Rust has raw strings with a variable count of `#`... 

I do not think we can find a sentinel value that can't appear in the input data but can be used in the column names simply because both, input data and the output can be strings. Hence, the domain of the symbols in the input can be the same as the domain of the symbols in the output -> no sentinel value may exist (assuming independence of the symbols).

Requirements for the naming function:
1. It must be reasonably fast.
2. The behavior of the function must be described in the documentation. E.g.:
```
    use_cat_names: bool
        if True, category values will be included in the encoded column names. Since this can result into duplicate column names, duplicates are suffixed with '#' symbol until a unique name is generated. If False, category indices will be used instead of the category values.
```

But I do not care about which symbol(s) we use and where do we place them.

Preservation of the attribute order should be another issue. #132 is already nontrivial.

**Comment by JohnnyC08** [comment]
Dang it you're right. Ok, I'm going to focus on the suffix `#` approach

**Comment by JohnnyC08** [comment]
@janmotl 
Went ahead and went with the suffix approach. Tell me what you think. Thanks buddy.

**Comment by JohnnyC08** [comment]
Ok, I have fixed the inverse transforming by creating a mapping object that gets used in get_dummies and reverse_dummies. That way we don't generate the column names at transform time, we generate them at fit time which I think makes sense. 

I tried to find opportunities to use the vectorization transforms we've started using in target encoder and the woe encoder, but I had trouble seeing the opportunities. Can you spot where I can use them?

Tell me what you think.

**Comment by janmotl** [comment]
I am not sure this is going to be faster, but for each nominal column:
1. Let's have a dictionary from category -> OHE column index. Pandas Series can be used for that as well.
2. Let's prealocate output matrix with zeros.
3. Then we can just loop over the nominal column once, read the category value, look up the index, write at the appropriate row and column index `1`. This step can be vectorized.
4. Append the generated matrix (which is just for this column) to the result.

Alternative: Instead of having the dictionary, we can have an encoding matrix. And instead of writing ones into the prealocated table, we could write whole rows from the encoding matrix.

Either way, the advantage is that we go thru the nominal column only once and not as many times as there is unique values in the column.

But before doing something, it could be meaningful to benchmark other implementations like `pandas.get_dummies` and see how they do it.
